File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/Knapsack.java
package com.example.dp;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class Knapsack {
    private Knapsack() {
    }

    public static Result memoized(int[] weights, int[] values, int capacity) {
        validate(weights, values, capacity);
        Map<String, Integer> memo = new HashMap<>();
        int best = solve(weights, values, 0, capacity, memo);
        return new Result("0/1 Knapsack", "memoization", best, Map.of());
    }

    private static int solve(int[] weights, int[] values, int index, int remaining, Map<String, Integer> memo) {
        if (index == weights.length || remaining == 0) {
            return 0;
        }
        String key = index + ":" + remaining;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        int skip = solve(weights, values, index + 1, remaining, memo);
        int take = weights[index] <= remaining
                ? values[index] + solve(weights, values, index + 1, remaining - weights[index], memo)
                : 0;
        int best = Math.max(skip, take);
        memo.put(key, best);
        return best;
    }

    public static Result tabulated(int[] weights, int[] values, int capacity) {
        validate(weights, values, capacity);
        int[][] dp = new int[weights.length + 1][capacity + 1];
        for (int i = 1; i <= weights.length; i++) {
            for (int cap = 0; cap <= capacity; cap++) {
                dp[i][cap] = dp[i - 1][cap];
                if (weights[i - 1] <= cap) {
                    dp[i][cap] = Math.max(dp[i][cap], values[i - 1] + dp[i - 1][cap - weights[i - 1]]);
                }
            }
        }
        List<Integer> items = new ArrayList<>();
        int cap = capacity;
        for (int i = weights.length; i > 0; i--) {
            if (dp[i][cap] != dp[i - 1][cap]) {
                items.add(i - 1);
                cap -= weights[i - 1];
            }
        }
        Collections.reverse(items);
        return new Result("0/1 Knapsack", "tabulation", dp[weights.length][capacity], Map.of("items", items));
    }

    public static Result spaceOptimized(int[] weights, int[] values, int capacity) {
        validate(weights, values, capacity);
        int[] dp = new int[capacity + 1];
        for (int i = 0; i < weights.length; i++) {
            for (int cap = capacity; cap >= weights[i]; cap--) {
                dp[cap] = Math.max(dp[cap], values[i] + dp[cap - weights[i]]);
            }
        }
        return new Result("0/1 Knapsack", "space optimized", dp[capacity], Map.of());
    }

    private static void validate(int[] weights, int[] values, int capacity) {
        if (capacity < 0 || weights.length != values.length) {
            throw new IllegalArgumentException("invalid knapsack input");
        }
    }
}

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/knapsack.py
from __future__ import annotations

from functools import lru_cache

from dp_algorithms.common import DPResult, require_non_negative


def _validate(weights: list[int], values: list[int], capacity: int) -> None:
    require_non_negative(capacity, "capacity")
    if len(weights) != len(values):
        raise ValueError("weights and values must have the same length")
    if any(weight < 0 for weight in weights):
        raise ValueError("weights must be non-negative")


def memoized(weights: list[int], values: list[int], capacity: int) -> DPResult:
    _validate(weights, values, capacity)

    @lru_cache(maxsize=None)
    def solve(index: int, remaining: int) -> int:
        if index == len(weights) or remaining == 0:
            return 0
        skip = solve(index + 1, remaining)
        take = 0
        if weights[index] <= remaining:
            take = values[index] + solve(index + 1, remaining - weights[index])
        return max(skip, take)

    return DPResult("0/1 Knapsack", "memoization", solve(0, capacity))


def tabulated(weights: list[int], values: list[int], capacity: int) -> DPResult:
    _validate(weights, values, capacity)
    n = len(weights)
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        weight = weights[i - 1]
        value = values[i - 1]
        for cap in range(capacity + 1):
            dp[i][cap] = dp[i - 1][cap]
            if weight <= cap:
                dp[i][cap] = max(dp[i][cap], value + dp[i - 1][cap - weight])

    chosen: list[int] = []
    cap = capacity
    for i in range(n, 0, -1):
        if dp[i][cap] != dp[i - 1][cap]:
            chosen.append(i - 1)
            cap -= weights[i - 1]
    chosen.reverse()
    return DPResult("0/1 Knapsack", "tabulation", dp[n][capacity], {"items": chosen})


def space_optimized(weights: list[int], values: list[int], capacity: int) -> DPResult:
    _validate(weights, values, capacity)
    dp = [0] * (capacity + 1)
    for weight, value in zip(weights, values, strict=True):
        for cap in range(capacity, weight - 1, -1):
            dp[cap] = max(dp[cap], value + dp[cap - weight])
    return DPResult("0/1 Knapsack", "space optimized", dp[capacity])
